File: go/example/sort/service/quicksortservice/pure/quicksortservice.go
package pure

import (
    otteropquicksortservice "github.com/otterop/example-sort/go/example/sort/service/quicksortservice"
    otteroparray "github.com/otterop/otterop/go/lang/array"
    otteropstring "github.com/otterop/otterop/go/lang/string"
)

type QuicksortService struct {
    otterop *otteropquicksortservice.QuicksortService
}


func Wrap(otterop *otteropquicksortservice.QuicksortService) *QuicksortService {
    ret := new(QuicksortService)
    ret.otterop = otterop
    return ret
}


func NewQuicksortService() *QuicksortService {
    ret := new(QuicksortService)
    ret.otterop = otteropquicksortservice.NewQuicksortService()
    return ret
}



func (this *QuicksortService) Sort(array []*string) []*string {
    var _arrayArray = make([]*otteropstring.String, len(array))
    for i := 0; i < len(array); i++ {
        _arrayArray[i] = otteropstring.Wrap(array[i])
    }
    var _array = otteroparray.Wrap(_arrayArray);
    retOtterop := this.otterop.Sort(_array)
    ret := make([]*string, retOtterop.Size())
    for i := 0; i < retOtterop.Size(); i++ {
        retI := retOtterop.Get(i)
        ret[i] = retI.Unwrap()
    }
    return ret
}

func (this *QuicksortService) Unwrap() *otteropquicksortservice.QuicksortService {
    return this.otterop
}

File: ts/example/sort/service/pure/QuicksortService.ts
import { String as StringOtterOP } from '@otterop/lang/string';
import { Array as ArrayOtterOP } from '@otterop/lang/array';
import { QuicksortService as QuicksortServiceOtterOP } from '../QuicksortService';
import { SortService } from './SortService';

export class QuicksortService implements SortService {

    private otterop : QuicksortServiceOtterOP;

    public static wrap(otterop :QuicksortServiceOtterOP) : QuicksortService {
        return new QuicksortService(otterop);
    }

    public constructor();
    public constructor(wrapped : QuicksortServiceOtterOP);
    public constructor(...__args: any[]) {
        if (__args.length === 1 && __args[0] instanceof QuicksortServiceOtterOP) {
            this.otterop = __args[0];
            return;
        }
        this.otterop = new QuicksortServiceOtterOP();
    }

    public sort(array : string[]) : string[] {
        const _arrayArray = new Array(array.length);
        for (let i = 0; i < array.length; i++) {
            _arrayArray[i] = StringOtterOP.wrap(array[i])
        }
        const _array = ArrayOtterOP.wrap(_arrayArray);
        const retOtterop = this.otterop.sort(_array);
        const ret = new Array(retOtterop.size());
        for (let i = 0; i < ret.length; i++) {
            const retI = retOtterop.get(i);
            ret[i] = retI.unwrap();
        }
        return ret;
    }

    public unwrap() : QuicksortServiceOtterOP {
        return this.otterop;
    }
}

